notebooks/e2e/e2e.py

```diff
@@ -0,0 +1,921 @@
+
+# split out into separate configs to minimise line length
+uc_config = config["urban_centre"]
+pop_config = config["population"]
+gtfs_config = config["gtfs"]
+osm_config = config["osm"]
+analyse_net_config = config["analyse_network"]
+metrics_config = config["metrics"]
+
+# %% [markdown] noqa: D212, D400, D415
+"""
+## Urban Centre Detection
+
+Merge 1Km gridded data together. Then detect the urban centre.
+
+### Data Sources
+
+Using [GHS-POP 1Km gridded](https://ghsl.jrc.ec.europa.eu/download.php?ds=pop)
+population estimaes, in a **Mollweide CRS**. The following tiles are expected
+in `config["urban_centre"]["input_dir"]`(which include the British isles and
+France):
+
+- R3-C18
+- R3-C19
+- R4-C18
+- R4-C19
+"""
+
+# %%
+# merge the urban centre input raster files to form one larger area
+# use subset regex to ensure expected year, CRS and resolution are used
+if uc_config["override"]:
+    merge_raster_files(
+        here(uc_config["input_dir"]),
+        os.path.dirname(here(uc_config["merged_path"])),
+        os.path.basename(uc_config["merged_path"]),
+        subset_regex="GHS_POP_E2020_GLOBE_R2023A_54009_1000_",
+    )
+
+# %%
+# put bbox into a geopandas dataframe for `get_urban_centre` input
+bbox_gdf = gpd.GeoDataFrame(
+    geometry=[box(*uc_config["bbox"])], crs="ESRI:54009"
+)
+
+# detect urban centre
+uc = UrbanCentre(here(uc_config["merged_path"]))
+uc_gdf = uc.get_urban_centre(
+    bbox_gdf,
+    centre=tuple(uc_config["centre"]),
+    buffer_size=uc_config["buffer_size"],
+)
+
+# set the index to the label column to make filtering easier
+uc_gdf.set_index("label", inplace=True)
+# %%
+# visualise outputs
+m = uc_gdf[::-1].reset_index().explore("label", cmap="viridis")
+
+# write to file
+if uc_config["write_outputs"]:
+    if not os.path.exists(os.path.dirname(here(uc_config["output_map_path"]))):
+        os.makedirs(os.path.dirname(here(uc_config["output_map_path"])))
+    m.save(here(uc_config["output_map_path"]))
+
+m
+# %% [markdown] noqa: D212, D400, D415
+"""
+## Population
+
+Merge 100m gridded data sources together, then resample onto a 200m grid by
+summing the consituent cells. Then retrive population data within the
+buffered urban centre boundary detected in the step above.
+
+### Data Sources
+
+Using [GHS-POP 100m gridded](https://ghsl.jrc.ec.europa.eu/download.php?ds=pop)
+population estimaes, in a **Mollweide CRS**. The following tiles are expected
+in `config["population"]["input_dir"]`(which include the British isles and
+France):
+
+- R3-C18
+- R3-C19
+- R4-C18
+- R4-C19
+"""
+
+# %%
+# merge the population input raster files to form one larger area
+# use regex to ensure 2020 data, CRS, and resolution are as expected in name
+if pop_config["override"]:
+    merge_raster_files(
+        here(pop_config["input_dir"]),
+        os.path.dirname(here(pop_config["merged_path"])),
+        os.path.basename(pop_config["merged_path"]),
+        subset_regex="GHS_POP_E2020_GLOBE_R2023A_54009_100_",
+    )
```

Comment:
`merge_raster_files` uses `check_parent_dir_exists(create=False)` - change to True or expose this parameter in the api for `merge_raster_files`, running in the e2e notebook required me to create the directory. Suggest raising separate issue.

Reply:
Setting `create=True` here is not the right strategy, it will lead to incorrect and unintuitive function behaviour. This function expects input files within an existing directory, and it's this directory we're checking. Creating the directory if it doesn't exist wouldn't work here since if the directory doesn't exist the input files do not exist either. Therefore, creating a directory if it doesn't exist would just lead to an error a few lines later when the search for .tif files begin. It could also lead to lint if a user forgets to delete the empty folder. I'd recommend to not raise the separate issue.

This function has however been fully brought up to date with `defence.py`in PR #146.

Reply:
I take your point here. Unfortunately, running the notebook results in `FileNotFoundError: Parent directory ~/Users/richardleyshon/Documents/transport-network-performance/data/external/population ~ not found on disk.`

If you would prefer not to expose the parameter, then it would be best to ensure the directory can be depended upon to exist. Chucking a .gitkeep in the appropriate folder and updating the .gitignore behaviour, for example.

Reply:
This is the expected behaviour - the error you are getting is the exact error that needs to be thrown. If the input directory doesn't exist then an error is needed since it can't proceed without any input data.